File: ognibuild/vcs.py
#!/usr/bin/python3
# Copyright (C) 2020 Jelmer Vernooij <jelmer@jelmer.uk>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

import errno

from breezy.errors import NotBranchError
from breezy.export import export
from breezy.workingtree import WorkingTree

from buildlog_consultant.sbuild import (
    NoSpaceOnDevice,
)

from . import DetailedFailure


def export_vcs_tree(tree, directory, subpath=""):
    try:
        export(tree, directory, "dir", None, subdir=(subpath or None))
    except OSError as e:
        if e.errno == errno.ENOSPC:
            raise DetailedFailure(1, ["export"], NoSpaceOnDevice()) from e
        raise


def dupe_vcs_tree(tree, directory):
    with tree.lock_read():
        if isinstance(tree, WorkingTree):
            tree = tree.basis_tree()
    try:
        result = tree._repository.controldir.sprout(
            directory, create_tree_if_local=True,
            revision_id=tree.get_revision_id()
        )
    except OSError as e:
        if e.errno == errno.ENOSPC:
            raise DetailedFailure(1, ["sprout"], NoSpaceOnDevice()) from e
        raise
    if not result.has_workingtree():
        raise AssertionError
    # Copy parent location - some scripts need this
    if isinstance(tree, WorkingTree):
        parent = tree.branch.get_parent()
    else:
        try:
            parent = tree._repository.controldir.open_branch().get_parent()
        except NotBranchError:
            parent = None
    if parent:
        result.open_branch().set_parent(parent)

File: src/vcs.rs
//! VCS-related functions
use breezyshim::branch::Branch;
use breezyshim::error::Error as BrzError;
use breezyshim::prelude::Repository;
use breezyshim::tree::PyTree;
use breezyshim::workingtree::{GenericWorkingTree, WorkingTree};
use std::path::{Path, PathBuf};
use url::Url;

/// Export a VCS tree to a new location.
///
/// # Arguments
/// * `tree` - The tree to export
/// * `directory` - The directory to export the tree to
/// * `subpath` - The subpath to export
pub fn export_vcs_tree<T: PyTree>(
    tree: &T,
    directory: &Path,
    subpath: Option<&Path>,
) -> Result<(), BrzError> {
    breezyshim::export::export(tree, directory, subpath)
}

/// A Breezy tree that can be duplicated.
pub trait DupableTree {
    /// Get the basis tree of this tree.
    fn basis_tree(&self) -> breezyshim::tree::RevisionTree;

    /// Get the parent location of this tree.
    fn get_parent(&self) -> Option<String>;

    /// Get the base directory of this tree, if it has one.
    fn basedir(&self) -> Option<PathBuf>;

    /// Export this tree to a directory.
    fn export_to(&self, directory: &Path, subpath: Option<&Path>) -> Result<(), BrzError>;
}

impl DupableTree for GenericWorkingTree {
    fn basis_tree(&self) -> breezyshim::tree::RevisionTree {
        WorkingTree::basis_tree(self).unwrap()
    }

    fn get_parent(&self) -> Option<String> {
        WorkingTree::branch(self).get_parent()
    }

    fn basedir(&self) -> Option<PathBuf> {
        Some(WorkingTree::basedir(self))
    }

    fn export_to(&self, directory: &Path, subpath: Option<&Path>) -> Result<(), BrzError> {
        export_vcs_tree(self, directory, subpath)
    }
}

impl DupableTree for breezyshim::tree::RevisionTree {
    fn basis_tree(&self) -> breezyshim::tree::RevisionTree {
        self.repository()
            .revision_tree(&self.get_revision_id())
            .unwrap()
    }

    fn get_parent(&self) -> Option<String> {
        let branch = self.repository().controldir().open_branch(None).unwrap();

        branch.get_parent()
    }

    fn basedir(&self) -> Option<PathBuf> {
        None
    }

    fn export_to(&self, directory: &Path, subpath: Option<&Path>) -> Result<(), BrzError> {
        export_vcs_tree(self, directory, subpath)
    }
}

/// Duplicate a VCS tree to a new location, including all history.
///
/// For a RevisionTree, this will duplicate the tree to a new location.
/// For a WorkingTree, this will duplicate the basis tree to a new location.
///
/// # Arguments
/// * `orig_tree` - The tree to duplicate
/// * `directory` - The directory to duplicate the tree to
pub fn dupe_vcs_tree(orig_tree: &dyn DupableTree, directory: &Path) -> Result<(), BrzError> {
    let tree = orig_tree.basis_tree();
    let result = tree.repository().controldir().sprout(
        Url::from_directory_path(directory).unwrap(),
        None,
        Some(true),
        None,
        Some(&tree.get_revision_id()),
    )?;

    assert!(result.has_workingtree());

    // Copy parent location - some scripts need this
    if let Some(parent) = orig_tree.get_parent() {
        let mut branch = result.open_branch(None)?;
        branch.set_parent(&parent);
    }

    Ok(())
}
